app/api/api/filters.py

```diff
@@ -0,0 +1,145 @@
+from django.db.models.query_utils import Q
+from rest_framework import filters
+from shared.data.models import VisibilityChoices
+
+
+class ScanReportAccessFilter(filters.BaseFilterBackend):
+    """
+    Filter that only allows users to see Scan Reports they are allowed to view, edit, or admin.
+
+    """
+
+    # Each model type needs a different relationship to get the Scan Report / Dataset permissions.
+    RELATIONSHIP_MAPPING = {
+        "scanreport": "",
+        "scanreporttable": "scan_report__",
+        "scanreportfield": "scan_report_table__scan_report__",
+        "scanreportvalue": "scan_report_field__scan_report_table__scan_report__",
+    }
+
+    def filter_queryset(self, request, queryset, view):
+        """
+        Filters the queryset based on visibility and permissions related to datasets and scan reports.
+
+        Args:
+            - request: The request object.
+            - queryset: The queryset to be filtered based on visibility and permissions.
+
+        Returns:
+            - A filtered queryset based on the specified visibility and permission conditions.
+        """
+
+        model = queryset.model.__name__.lower()
+        relationship = self.RELATIONSHIP_MAPPING.get(model, "")
+
+        dataset_visibility = f"{relationship}parent_dataset__visibility"
+        scan_report_visibility = f"{relationship}visibility"
+        scan_report_viewers = f"{relationship}viewers"
+        scan_report_editors = f"{relationship}editors"
+        scan_report_author = f"{relationship}author"
+        dataset_viewers = f"{relationship}parent_dataset__viewers"
+        dataset_editors = f"{relationship}parent_dataset__editors"
+        dataset_admins = f"{relationship}parent_dataset__admins"
+
+        return queryset.filter(
```

Comment:
I haven't looked at the code for too long to be sure this isn't implemented elsewhere: but I think here we also need to implement Project-level filtering. Users should never be able to see a SR that isn't in one of the Projects they are a member of, even if that SR and its Dataset are set as "PUBLIC".

The only exception is AZ_FUNCTION_USER who can see everything to allow mapping reuse.

Reply:
Thanks - hadn't thought of the project level, I will add this.

Reply:
good spot @spco